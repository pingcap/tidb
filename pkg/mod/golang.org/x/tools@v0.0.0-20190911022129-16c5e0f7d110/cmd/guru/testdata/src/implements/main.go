package main

// Tests of 'implements' query.
// See go.tools/guru/guru_test.go for explanation.
// See implements.golden for expected query results.

import _ "lib"

func main() {
}

type E interface{} // @implements E "E"

type F interface { // @implements F "F"
	f()
}

type FG interface { // @implements FG "FG"
	f()
	g() []int // @implements slice "..int"
}

type C int // @implements C "C"
type D struct{}

func (c *C) f() {} // @implements starC ".C"
func (d D) f()  {} // @implements D "D"

func (d *D) g() []int { return nil } // @implements starD ".D"

type sorter []int // @implements sorter "sorter"

func (sorter) Len() int           { return 0 }
func (sorter) Less(i, j int) bool { return false }
func (sorter) Swap(i, j int)      {}

type I interface { // @implements I "I"
	Method(*int) *int
}

func _() {
	var d D
	_ = d // @implements var_d "d"
}
