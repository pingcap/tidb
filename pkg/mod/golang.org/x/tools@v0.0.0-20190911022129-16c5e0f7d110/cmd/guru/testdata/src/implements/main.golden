-------- @implements E --------
empty interface type E

-------- @implements F --------
interface type F
	is implemented by pointer type *C
	is implemented by struct type D
	is implemented by interface type FG

-------- @implements FG --------
interface type FG
	is implemented by pointer type *D
	implements F

-------- @implements slice --------
slice type []int implements only interface{}

-------- @implements C --------
pointer type *C
	implements F

-------- @implements starC --------
pointer type *C
	implements F

-------- @implements D --------
struct type D
	implements F
pointer type *D
	implements FG

-------- @implements starD --------
pointer type *D
	implements F
	implements FG

-------- @implements sorter --------
slice type sorter
	implements lib.Sorter

-------- @implements I --------
interface type I
	is implemented by basic type lib.Type

-------- @implements var_d --------
struct type D
	implements F
pointer type *D
	implements FG

